{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "290fa236",
   "metadata": {},
   "source": [
    "# Introduction to Python Programming for Biology\n",
    "\n",
    "This notebook introduces basic Python concepts using simple biology-related examples."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fd946b21",
   "metadata": {},
   "source": [
    "## Part 1: Hello World ðŸ‘‹"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b2e4a62d",
   "metadata": {},
   "source": [
    "The traditional first program in any language prints the words \"Hello, world!\".\n",
    "\n",
    "**Example:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ccce09",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Hello, world!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2609e589",
   "metadata": {},
   "source": [
    "### Exercise\n",
    "Modify the code above to print your name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7c90a2b",
   "metadata": {},
   "source": [
    "## Part 2: Learning Functions"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "85bc5459",
   "metadata": {},
   "source": [
    "A function is a reusable block of code that performs a specific task.\n",
    "\n",
    "**Example: A simple function**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd639450",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def greet(name):\n",
    "    print(\"Hello\", name)\n",
    "\n",
    "# call the function\n",
    "greet(\"Alice\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e7e135d8",
   "metadata": {},
   "source": [
    "### Using conditions inside functions\n",
    "\n",
    "We can change what a function prints depending on the input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "072eb6d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def describe_mood(mood):\n",
    "    if mood == \"happy\":\n",
    "        print(\"I am feeling happy today!\")\n",
    "    elif mood == \"sad\":\n",
    "        print(\"I am feeling a bit sad today.\")\n",
    "    else:\n",
    "        print(\"I am feeling okay.\")\n",
    "\n",
    "describe_mood(\"happy\")\n",
    "describe_mood(\"sad\")\n",
    "describe_mood(\"excited\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88b15aac",
   "metadata": {},
   "source": [
    "### Exercise\n",
    "Write a function called `get_organism_type` that prints:\n",
    "\n",
    "* \"This is a bacterium\" if input is `\"bacteria\"`\n",
    "\n",
    "* \"This is a virus\" if input is `\"virus\"`\n",
    "\n",
    "* \"Unknown organism\" otherwise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78a63251",
   "metadata": {},
   "source": [
    "## Part 3: Reading and Writing DNA Sequences"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0c0e9f51",
   "metadata": {},
   "source": [
    "DNA sequences can be stored as strings in Python.\n",
    "\n",
    "**Example: DNA Sequence as a string**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8659184",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dna = \"ATGCGTACGTTAG\"\n",
    "print(dna)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0359a5fd",
   "metadata": {},
   "source": [
    "### Reading a DNA sequence from a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d36b14a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(\"dna.txt\", \"w\") as file:\n",
    "    file.write(\"ATGAAATTTGGG\")\n",
    "\n",
    "with open(\"dna.txt\", \"r\") as file:\n",
    "    dna = file.read().strip()\n",
    "\n",
    "print(dna)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "94a9a90e",
   "metadata": {},
   "source": [
    "### Writing a DNA sequence to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb538b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_dna = \"ATGAAATTTGGG\"\n",
    "\n",
    "\n",
    "with open(\"output_dna.txt\", \"w\") as file:\n",
    "    file.write(new_dna)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3b364c6a",
   "metadata": {},
   "source": [
    "**Exercise**: Write the sentence **'I enjoy Python programming'** to a file named `like_python.txt`. Read the sentence back and print it out."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "070e1c99",
   "metadata": {},
   "source": [
    "### Better ways of handling biological sequences\n",
    "\n",
    "The exercise above illustrates an issue of reading/writing biological sequences - this can work on anything, and is not specific for biological sequences.\n",
    "\n",
    "In Python there are *packages* specifically designed for handling different operations and analysis related to biological data.\n",
    "\n",
    "In the lecture we introduced the FASTA format for storing biological sequences - a python package named Biopython has functionalities for defining DNA and amino acid sequences, as well as reading/writing sequences in the FASTA format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62747062",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run once if Biopython is not installed\n",
    "# in a jupyter notebok, a '!' at the front of the line means this is to be run on the terminal on the system, rather than a Python command\n",
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "634bce22",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fd5a39eb",
   "metadata": {},
   "source": [
    "### Defining a DNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddcd5405",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_seq = Seq(\"ATGCGTACGTTAG\")\n",
    "dna_seq\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "db03999e",
   "metadata": {},
   "source": [
    "### Create a SeqRecord\n",
    "\n",
    "A `SeqRecord` also allows you to define identifiers and description for the sequence. (`SeqRecord` and `Seq` are examples of Python objects [they are technically referred to as 'classes' in Python] - recall discussion on object-oriented programming in the lecture.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af82e340",
   "metadata": {},
   "outputs": [],
   "source": [
    "record = SeqRecord(\n",
    "    dna_seq,\n",
    "    id=\"example_dna\",\n",
    "    description=\"A short example DNA sequence\"\n",
    ")\n",
    "\n",
    "record\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "72aa03cb",
   "metadata": {},
   "source": [
    "### Write the DNA sequence to a FASTA file"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c6970395",
   "metadata": {},
   "source": [
    "`SeqIO` is the utility from Biopython that allow you to perform read/write operations of sequences - 'IO' refers to input/output, which is one of the most basic operations of data one can perform in computer programming!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47b4d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"example_dna.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(record, output_handle, \"fasta\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e2f16a06",
   "metadata": {},
   "source": [
    "Try opening the `example_dna.fasta` file in text format and remind yourself of the requirements of the FASTA format."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a32eb63f",
   "metadata": {},
   "source": [
    "### Read the DNA sequence back from the file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de9da4c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"example_dna.fasta\", \"r\") as input_handle:\n",
    "    records = list(SeqIO.parse(input_handle, \"fasta\"))\n",
    "\n",
    "records\n",
    "\n",
    "# Acess the sequence\n",
    "dna_from_file = records[0].seq\n",
    "dna_from_file\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3137ae68",
   "metadata": {},
   "source": [
    "Another advantage of manipulating sequences with Biopython classes is that the class comes with functions allowing you to perform useful operations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88f43d8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Length of the sequence\n",
    "len(dna_from_file)\n",
    "\n",
    "# Reverse complement\n",
    "dna_from_file.reverse_complement()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "055635c3",
   "metadata": {},
   "source": [
    "## Part 4: Translating DNA to Amino Acids ðŸ§¬"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "03cc075b",
   "metadata": {},
   "source": [
    "Here we will practice writing simple algorithms in Python. As an example we will consider the problem of translation: we know that DNA is translated into amino acids using **codons** (groups of three nucleotides)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c70f2334",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "codon_table = {\n",
    "    \"ATG\": \"M\", # Methionine (start codon)\n",
    "    \"TTT\": \"F\",\n",
    "    \"TTC\": \"F\",\n",
    "    \"TAA\": \"*\", # Stop codon\n",
    "    \"TAG\": \"*\",\n",
    "    \"TGA\": \"*\"\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "edf8bc4b",
   "metadata": {},
   "source": [
    "*(Note: This is a simplified table just for illustration - we know of course the codon table is much larger than this!)*"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "48632c0f",
   "metadata": {},
   "source": [
    "At the end of this part we would like to write a simple algorithm that takes in a DNA sequence as input, and return the translated DNA sequence following a codon table. \n",
    "\n",
    "This is possible with functionalities available in Biopython - but just for teaching & learning purposes we will break the problem down step by step, and think about how to implement this using basic operation in Python without specialised packages."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d97bb853",
   "metadata": {},
   "source": [
    "### Some related tools we need for this\n",
    "\n",
    "#### 'Looping' over lists\n",
    "\n",
    "One thing we need to learn before approaching the problem is how to efficiently perform repeated oeprations over a bunch of data entries. In programming we call this 'looping'. \n",
    "\n",
    "A simple loop is the 'for' loop:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "309cfe1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "animals = [\"cat\", \"dog\", \"mouse\"]\n",
    "\n",
    "for animal in animals:\n",
    "    print(animal)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "caafbedd",
   "metadata": {},
   "source": [
    "You can also loop over a list of numbers:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f942c0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to define a list of numbers we can use the python function 'range'\n",
    "print(range(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5733b794",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(5):\n",
    "    print(i) # each entry in the list of numbers printed in a new line\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "507c41c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# By default it starts from '0'. It doesn't have to be the case:\n",
    "for i in range(1, 10):\n",
    "    print(i) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2fa4756",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Also, we don't have to iterate step-by-step.\n",
    "for i in range(0, 10, 2):\n",
    "    print(i)\n",
    "\n",
    "# function 'range' expects: range(start, stop, step)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ef896be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# going back to the example of strings. A string is actually a list of letters in Python. Try this out:\n",
    "for i in 'I like Python programming':\n",
    "    print(i) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acf1ec4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we can also use a for loop to generate a new string!\n",
    "my_string = 'I like Python programming'\n",
    "print(my_string)\n",
    "for i in ['because', 'it', 'is', 'fun']:\n",
    "    j = ' ' + i # '+' here means concatenation (i.e. sticking them together). Just putting a space in front as it looks nicer\n",
    "    my_string += ' ' + i # '+=' is a convenient way to do BOTH concatenation and assignment (e.g. y += x means the same as y = y + x)\n",
    "    print(my_string)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "981c9d5e",
   "metadata": {},
   "source": [
    "**Exercise**: write a 'for' loop that prints an input DNA sequence in triplets. (Let's use as example the input string `\"ATGTTTTAA\"`)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ca622951",
   "metadata": {},
   "source": [
    "### Python 'dictionary'\n",
    "\n",
    "In python, 'dictionary' is a data structure that maps 'keys' to their corresponding 'values'. Think of it like a real dictionary: *word* maps to *definition*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99fdfdae",
   "metadata": {},
   "outputs": [],
   "source": [
    "ages = {\n",
    "    \"Alice\": 20,\n",
    "    \"Bob\": 25,\n",
    "    \"Charlie\": 30\n",
    "}\n",
    "\n",
    "print(ages)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "84d805d3",
   "metadata": {},
   "source": [
    "The advantage of the key-value structure is that you can look up the value corresponding to a key (just like you look up definitions of a word in dictionary):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f4ee8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ages[\"Alice\"])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "92fcde75",
   "metadata": {},
   "source": [
    "A requirement is that the keys have to be unique!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce697698",
   "metadata": {},
   "outputs": [],
   "source": [
    "example = {\n",
    "    \"A\": 1,\n",
    "    \"A\": 2\n",
    "}\n",
    "\n",
    "print(example) # the second value overwrites the first!!!\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "df30d933",
   "metadata": {},
   "source": [
    "A safe way to perfofrm this lookup is to use the *dictionary method* `.get()`.\n",
    "* If key exists â†’ return its value\n",
    "* If key does not exist â†’ return `default_value`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf918bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ages.get(\"Bob\", \"N/A\")) # should return a value ('Bob' is one of the keys in the dictionary!)\n",
    "print(ages.get(\"Dave\", \"N/A\")) # but 'Dave' isn't ...\n",
    "print(ages.get(\"Dave\", \"I don't know\")) # How about this?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "b646c179",
   "metadata": {},
   "source": [
    "**Exercise:** Create a dictionary that function as the table of triplet DNA codes you learnt in basic molecular biology, and verify that you can query the dictionary to return an amino acid given input triplets.\n",
    "\n",
    "(You might want to search for the codon table - this is a classic Biocomputing exercise, e.g. https://gist.github.com/juanfal/09d7fb53bd367742127e17284b9c47bf)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fcd976b2",
   "metadata": {},
   "source": [
    "### Final Exercise\n",
    "\n",
    "Write the complete algorithm! Here are some helpful hints:\n",
    "\n",
    "* you want the program to stop when it sees a stop codon\n",
    "* make sure it is capable of handling 'wrong' DNA triplets (otherwise it will break the for loop ...)\n",
    "* remember a string can be thought of as a list of letters. This applies to iterating through an existing string, but also is useful if you want to generate a new string (by repeatedly appending elements to the back of the string)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
